relayer/relayer.go

```diff
@@ -438,20 +450,67 @@ func (r *Relayer) RelayMessage(warpLogInfo *vmtypes.WarpLogInfo) error {
 	// Increment the request ID for the next message relay request
 	r.currentRequestID++
 
-	// Update the database with the latest processed block height
-	err = r.db.Put(r.sourceBlockchainID, []byte(database.LatestProcessedBlockKey), []byte(strconv.FormatUint(warpLogInfo.BlockNumber, 10)))
-	if err != nil {
-		r.logger.Error(
-			fmt.Sprintf("failed to put %s into database", database.LatestProcessedBlockKey),
-			zap.Error(err),
-		)
+	if !storeProcessedHeight {
+		return nil
 	}
 
-	return nil
+	// Update the database with the latest processed block height
+	return r.storeLatestBlockHeight(warpLogInfo.BlockNumber)
 }
 
 // Returns whether destinationBlockchainID is a supported destination.
 // If supportedDestinations is empty, then all destination chain IDs are supported.
 func (r *Relayer) CheckSupportedDestination(destinationBlockchainID ids.ID) bool {
 	return len(r.supportedDestinations) == 0 || r.supportedDestinations.Contains(destinationBlockchainID)
 }
+
+// Get the latest processed block height from the database.
+// Note that there may be unrelayed messages in the latest processed block
+// because it is updated as soon as a single message from that block is relayed,
```

Comment:
are there any plans to address this in the future? Think we'd need to somehow batch the Warp logs from the same block together somehow in this case.

Reply:
Nope. Lots of headache for very little benefit. Currently the worst-case scenario is that we reprocess one block containing transactions, which short-circuits if those messages have already been delivered.

Reply:
It's unfortunately a difficult problem to fully solve. The relayer processes Warp messages on a log-by-log basis, not a block-by-block basis, meaning that when the relayer sees a Warp message, it has no way of knowing if that's the last Warp message in that particular block.

So our choices are to either write the block to the DB as soon as we see it, or write it to the DB as soon as we stop seeing that block in new Warp messages. Since the time between Warp messages is unbounded, the former choice is the safest option. Therefore, when reading a block from the database on startup, all we know is that at least one Warp message from that block has been processed, but there may be more. The only workarounds I can think of would be to either process entire blocks atomically, or write each processed Warp message ID to the database, both of which are significant changes. OTOH, the only downside of the current approach is that messages may be processed more than once, but 1) that's limited to a single block and 2) there are safeguards in place to minimize the amount of work the relayer actually does in those cases.